// @stardust-collective/dag4-network/dist/cjs/api/v2/block-explorer-api.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.blockExplorerApi = exports.BlockExplorerV2Api = void 0;
const dag4_core_1 = require("@stardust-collective/dag4-core");
const DNC_1 = require("../../DNC");
class BlockExplorerV2Api {
    constructor(host) {
        this.service = new dag4_core_1.RestApi(DNC_1.DNC.BLOCK_EXPLORER_URL);
        if (host) {
            this.config().baseUrl(host);
        }
    }
    config() {
        return this.service.configure();
    }
    // Snapshots
    getSnapshot(hash) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.service.$get(`/global-snapshots/${hash}`);
        });
    }
    getTransactionsBySnapshot(hash, limit, searchAfter, searchBefore, next) {
        return __awaiter(this, void 0, void 0, function* () {
            const params = this.buildRequestParams({ limit, searchAfter, searchBefore, next });
            return this.service.$get(`/global-snapshots/${hash}/transactions`, params);
        });
    }
    getRewardsBySnapshot(hash, limit, next) {
        return __awaiter(this, void 0, void 0, function* () {
            const params = this.buildRequestParams({ limit, next });
            return this.service.$get(`/global-snapshots/${hash}/rewards`, params);
        });
    }
    getSnapshots(limit, searchAfter, searchBefore, next) {
        return __awaiter(this, void 0, void 0, function* () {
            const params = this.buildRequestParams({ limit, searchAfter, searchBefore, next });
            return this.service.$get(`/global-snapshots`, params);
        });
    }
    getLatestSnapshot() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.service.$get('/global-snapshots/latest');
        });
    }
    getLatestSnapshotTransactions(limit, searchAfter, searchBefore, next) {
        return __awaiter(this, void 0, void 0, function* () {
            const params = this.buildRequestParams({ limit, searchAfter, searchBefore, next });
            return this.service.$get(`/global-snapshots/latest/transactions`, params);
        });
    }
    getLatestSnapshotRewards(limit, next) {
        return __awaiter(this, void 0, void 0, function* () {
            const params = this.buildRequestParams({ limit, next });
            return this.service.$get(`/global-snapshots/latest/rewards`, params);
        });
    }
    // Private method
    buildRequestParams({ limit = null, searchAfter = null, searchBefore = null, next = null, actionType = null, active = false }) {
        let params;
        if (limit || searchAfter || searchBefore || next || active) {
            params = {};
            if (limit && limit > 0) {
                params.limit = limit;
            }
            if (actionType) {
                params.transactionType = actionType;
            }
            if (active) {
                params.active = active;
            }
            // search_after, search_before and next are mutually exclusive
            if (searchAfter) {
                params.search_after = searchAfter;
            }
            else if (searchBefore) {
                params.search_before = searchBefore;
            }
            else if (next) {
                params.next = next;
            }
        }
        return params;
    }
    // Transactions
    getTransactions(limit, searchAfter, searchBefore, next) {
        return __awaiter(this, void 0, void 0, function* () {
            const params = this.buildRequestParams({ limit, searchAfter, searchBefore, next });
            return this.service.$get(`/transactions`, params);
        });
    }
    getTransactionsByAddress(address, limit, searchAfter, sentOnly, receivedOnly, searchBefore, next) {
        return __awaiter(this, void 0, void 0, function* () {
            const searchPath = sentOnly ? '/sent' : receivedOnly ? '/received' : '';
            const params = this.buildRequestParams({ limit, searchAfter, searchBefore, next });
            return this.service.$get(`/addresses/${address}/transactions${searchPath}`, params);
        });
    }
    getTransaction(hash) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.service.$get(`/transactions/${hash}`);
        });
    }
    // Addresses
    getAddressBalance(address) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.service.$get(`/addresses/${address}/balance`);
        });
    }
    // Blocks
    getCheckpointBlock(hash) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.service.$get(`/blocks/${hash}`);
        });
    }
    // Metagraphs
    getLatestCurrencySnapshot(metagraphId) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.service.$get(`/currency/${metagraphId}/snapshots/latest`);
        });
    }
    getCurrencySnapshot(metagraphId, hashOrOrdinal) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.service.$get(`/currency/${metagraphId}/snapshots/${hashOrOrdinal}`);
        });
    }
    getLatestCurrencySnapshotRewards(metagraphId, limit, next) {
        return __awaiter(this, void 0, void 0, function* () {
            const params = this.buildRequestParams({ limit, next });
            return this.service.$get(`/currency/${metagraphId}/snapshots/latest/rewards`, params);
        });
    }
    getCurrencySnapshotRewards(metagraphId, hashOrOrdinal, limit, next) {
        return __awaiter(this, void 0, void 0, function* () {
            const params = this.buildRequestParams({ limit, next });
            return this.service.$get(`/currency/${metagraphId}/snapshots/${hashOrOrdinal}/rewards`, params);
        });
    }
    getCurrencyBlock(metagraphId, hash) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.service.$get(`/currency/${metagraphId}/blocks/${hash}`);
        });
    }
    getCurrencyAddressBalance(metagraphId, address) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.service.$get(`/currency/${metagraphId}/addresses/${address}/balance`);
        });
    }
    getCurrencyTransaction(metagraphId, hash) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.service.$get(`/currency/${metagraphId}/transactions/${hash}`);
        });
    }
    getCurrencyTransactions(metagraphId, limit, searchAfter, searchBefore, next) {
        return __awaiter(this, void 0, void 0, function* () {
            const params = this.buildRequestParams({ limit, searchAfter, searchBefore, next });
            return this.service.$get(`/currency/${metagraphId}/transactions`, params);
        });
    }
    getCurrencyTransactionsByAddress(metagraphId, address, limit, searchAfter, sentOnly, receivedOnly, searchBefore, next) {
        return __awaiter(this, void 0, void 0, function* () {
            const searchPath = sentOnly ? '/sent' : receivedOnly ? '/received' : '';
            const params = this.buildRequestParams({ limit, searchAfter, searchBefore, next });
            return this.service.$get(`/currency/${metagraphId}/addresses/${address}/transactions${searchPath}`, params);
        });
    }
    getCurrencyActionsByAddress(metagraphId, address, actionType, limit, searchAfter, searchBefore, next) {
        return __awaiter(this, void 0, void 0, function* () {
            const params = this.buildRequestParams({ limit, searchAfter, searchBefore, next, actionType });
            return this.service.$get(`/currency/${metagraphId}/addresses/${address}/actions`, params);
        });
    }
    getCurrencyTransactionsBySnapshot(metagraphId, hashOrOrdinal, limit, searchAfter, searchBefore, next) {
        return __awaiter(this, void 0, void 0, function* () {
            const params = this.buildRequestParams({ limit, searchAfter, searchBefore, next });
            return this.service.$get(`/currency/${metagraphId}/snapshots/${hashOrOrdinal}/transactions`, params);
        });
    }
    getCurrencyTokenLocksByAddress(metagraphId, address, limit, searchAfter, searchBefore, next, active) {
        return __awaiter(this, void 0, void 0, function* () {
            const params = this.buildRequestParams({ limit, searchAfter, searchBefore, next, active });
            return this.service.$get(`/currency/${metagraphId}/addresses/${address}/token-locks`, params);
        });
    }
    getCurrencyAllowSpendsByAddress(metagraphId, address, limit, searchAfter, searchBefore, next, active) {
        return __awaiter(this, void 0, void 0, function* () {
            const params = this.buildRequestParams({ limit, searchAfter, searchBefore, next, active });
            return this.service.$get(`/currency/${metagraphId}/addresses/${address}/allow-spends`, params);
        });
    }
    getActionsByAddress(address, actionType, limit, searchAfter, searchBefore, next) {
        return __awaiter(this, void 0, void 0, function* () {
            const params = this.buildRequestParams({ limit, searchAfter, searchBefore, next, actionType });
            return this.service.$get(`/addresses/${address}/actions`, params);
        });
    }
    getTokenLocksByAddress(address, limit, searchAfter, searchBefore, next, active) {
        return __awaiter(this, void 0, void 0, function* () {
            const params = this.buildRequestParams({ limit, searchAfter, searchBefore, next, active });
            return this.service.$get(`/addresses/${address}/token-locks`, params);
        });
    }
    getAllowSpendsByAddress(address, limit, searchAfter, searchBefore, next, active) {
        return __awaiter(this, void 0, void 0, function* () {
            const params = this.buildRequestParams({ limit, searchAfter, searchBefore, next, active });
            return this.service.$get(`/addresses/${address}/allow-spends`, params);
        });
    }
}
exports.BlockExplorerV2Api = BlockExplorerV2Api;
exports.blockExplorerApi = new BlockExplorerV2Api();
